pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,446 @@
+			} else {
+				glog.V(2).Infof("Unable to find the time to clean up Job %s/%s, dropping it: %v", job.Namespace, job.Name, err)
+			}
+			tc.popJobsDelayCleanupQueue(job)
+			continue
+		}
+		// If TTL expires, add the Job to the cleanup queue;
+		// otherwise, return because nothing left in the queue expires.
+		if *t <= 0 {
+			tc.popJobsDelayCleanupQueue(job)
+			tc.addJobToCleanup(job)
+		} else {
+			// This job is the top of the queue.
+			// Update the next check time to this job's expiration time.
+			tc.setNextCheckTo(expireAt)
+			return true
+		}
+	}
+}
+
+// popJobsDelayCleanupQueue is supposed to be called by handleJobsDelayCleanup only.
+func (tc *Controller) popJobsDelayCleanupQueue(top *batch.Job) {
+	// Pop from the jobs delay cleanup queue.
+	obj, err := tc.jobsDelayCleanupQueue.Pop()
+	if err != nil {
+		glog.V(4).Infof("Error: failed to pop from the delay cleanup queue: %v", err)
+	}
+	job, ok := obj.(*batch.Job)
+	if !ok {
+		glog.V(2).Infof("Error: popped obj from the delay cleanup queue, but it's not a Job: %#v", obj)
+	}
+	// If we popped something other than what we want to pop, we need to add it back.
+	// We expect this not to happen.
+	if job.Namespace != top.Namespace || job.Name != top.Name {
+		glog.V(2).Infof("Error: expect popped job from the delay cleanup queue to be the same as peeked one; popped: %#v, peeked: %#v", job, top)
+		tc.jobsDelayCleanupQueue.Add(job)
+		return
+	}
+	glog.V(4).Infof("Popped Job %s/%s from the delay cleanup queue", job.Namespace, job.Name)
+}
+
+// setNextCheckTo sets the nextCheckAt to the given time t
+func (tc *Controller) setNextCheckTo(t *time.Time) {
+	tc.nextCheckAtMutex.Lock()
+	defer tc.nextCheckAtMutex.Unlock()
+
+	if tc.nextCheckAt == nil {
+		tc.nextCheckAt = new(time.Time)
+	}
+	tUTC := t.UTC()
+	*tc.nextCheckAt = tUTC
+	glog.V(4).Infof("Next time to check jobs delay cleanup queue updated to: %v", *tc.nextCheckAt)
+}
+
+func (tc *Controller) handleErr(err error, key interface{}) {
+	if err == nil {
+		tc.jobsToCleanup.Forget(key)
+		return
+	}
+
+	utilruntime.HandleError(fmt.Errorf("error cleaning up Job %v, will retry: %v", key, err))
+	tc.jobsToCleanup.AddRateLimited(key)
+}
+
+// cleanupJob will check the Job's state and TTL and delete the Job when it
+// finishes and its TTL after finished has expired. This function is not meant
+// to be invoked concurrently with the same key.
+func (tc *Controller) cleanupJob(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Attempt to clean up Job %s", name)
+	// Ignore the Jobs that are already deleted or being deleted
+	job, err := tc.jLister.Jobs(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	if job.DeletionTimestamp != nil {
+		return nil
+	}
+
+	// Before deleting the Job, do a final sanity check just in case TTL is modified.
```

Comment:
I don't really see what this buys us over using the version in our watch-fed cache... the ttl could still change between this lookup and the delete that follows

Reply:
I am in between, caches can go stale, but as you say we don't guarantee the atomicity either. On the other hand the cost isn't high here - you do only live get if you were to call delete. 

Maybe we should make that field immutable?

Reply:
@liggitt @tnozicka The benefit is that we can be sure that the TTL has expired at this point; we can't be sure with the value from cache. As long as the TTL expires at some point, it's safe for the system to clean up the resource.  

For example, 

1. At t1, we expect TTL to expire at t1+10
1. TTL was changed and it will expire at t1+20, but we haven't seen it in cache.
1. At t1+10, we check if TTL actually expires, and find out it hasn't. If we don't do this, we can't be sure if TTL has expired at any moment - if TTL was changed before t1+10, the new value should be respected. 
1. At t2+20 we check again, TTL expires and it's safe to clean up now. 
1. Even if TTL was changed afterwards, you can't change the fact that the TTL has expired at some point and the resource can be cleaned up.

Reply:
@janetkuo I think you are correct.

Is there a use case to support changing the TTL after you set it? 

Also I guess for pods the spec is immutable anyways hence the TTL field - do we want to differ with JobSpec?

Reply:
> Is there a use case to support changing the TTL after you set it?

Some example use cases:
* The users have some existing Jobs that already complete, and they want to set their TTLs 
* The users create some Jobs without TTLs, and decide to set the TTLs after the Jobs are created (can be done with mutating webhooks). 
* The users create some Jobs with TTLs, and decide to unset/extend TTLs before the Jobs complete or before the TTL expire.

Making TTLs mutable is convenient for the users without sacrificing correctness (we can still tell whether a TTL has expired or not). At the very least we should allow TTLs to be shortened, but IMO it's safe to allow users to extend/unset TTLs and that would make it more useful for the users. 

> for pods the spec is immutable anyways

Pod's `image` field is mutable, so it's okay to make pod TTL mutable too.

Reply:
> The benefit is that we can be sure that the TTL has expired at this point; we can't be sure with the value from cache. As long as the TTL expires at some point, it's safe for the system to clean up the resource.

I can buy that, though it's worth documenting on the field that once the TTL has expired, the job can be deleted, even if an update to extend the TTL returns a successful API response.

Longer-term, we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

Reply:
> we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

This approach was discussed in the original proposal (in comments): https://goo.gl/YxtxTi
However the downside is that we'll have  a DELETE conflict whenever the resource (Job) is updated, but we actually only care about changes to TTL and nothing else. Moreover, DELETE doesn't return us the resource, so we'll need to GET the resource again (to get the latest TTL to re-add it to the queue) whenever there's a conflict.

Reply:
> However the downside is that we'll have a DELETE conflict whenever the resource (Job) is updated, but we actually only care about changes to TTL and nothing else. Moreover, DELETE doesn't return us the resource, so we'll need to GET the resource again (to get the latest TTL to re-add it to the queue) whenever there's a conflict.

As there is a conflict ("higher" RV) you can safely drop the item and wait for informers to catch up, no need for GET. I guess given the Job is suppose to be finished controllers don't really have much of a reason to update it and cause conflicts which leaves the person updating the TTL.

> Longer-term, we might want a resourceVersion delete precondition so you can issue a delete and count on the object not having changed out from under you

+1; I have needed that feature several times in the past